File: back/vscar-back-nestjs/src/cars/cars.controller.ts
import { Body, Controller, Get, Param, Post, Query } from '@nestjs/common';
import { PaginateResult } from 'mongoose';
import { CarsService } from './cars.service';
import { Car } from './schemas/car.schema';

@Controller('cars')
export class CarsController {
  constructor(private readonly carsService: CarsService) {}

  @Post('findByFilter')
  async findByFilter(@Body() jsonRequest): Promise<PaginateResult<Car>> {
    return this.carsService.findByFilter(jsonRequest);
  }

  @Get(':id')
  async findOne(@Param('id') id: number): Promise<Car> {
    return this.carsService.findOne(+id);
  }

  @Get()
  async findAll(): Promise<Car[]> {
    return this.carsService.findAll();
  }

  // TODO: move to another controller
  @Get('attributes/names')
  async findNames(@Query('text') text: string): Promise<any> {
    return this.carsService.findNames(text);
  }

  @Get('attributes/values')
  async findValues(@Query('text') text: string): Promise<any> {
    return this.carsService.findValues(text);
  }

  @Get('db/version')
  async getVersion(): Promise<string> {
    return this.carsService.getProcessedDate();
  }
}

File: back/vscar-back-spring/src/main/java/com/vscar/vscar_back_spring/CarsController.java
/* (C) 2025 */
package com.vscar.vscar_back_spring;

import com.vscar.vscar_back_spring.dto.Car;
import com.vscar.vscar_back_spring.dto.FilterQuery;
import com.vscar.vscar_back_spring.dto.PaginatedResult;
import java.util.List;
import java.util.Optional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/**
 * @author kiril
 */
@RestController
@RequestMapping("cars")
public class CarsController {

  private final CarsService carsService;

  @Autowired
  public CarsController(CarsService carsService) {
    this.carsService = carsService;
  }

  @PostMapping("findByFilter")
  public PaginatedResult findByFilter(@RequestBody FilterQuery query) {
    return this.carsService.findByFilter(query);
  }

  @GetMapping("{id}")
  public Car findOne(@PathVariable("id") Long id) {
    return this.carsService.findOne(id);
  }

  @GetMapping()
  public List<Car> findAll() {
    return this.carsService.findAll();
  }

  @GetMapping("attributes/names")
  public List<String> findNames(@RequestParam("text") Optional<String> text) {
    return this.carsService.findNames(text.orElse(""));
  }

  @GetMapping("attributes/values")
  public Object findValues(@RequestParam("text") Optional<String> text) {
    return this.carsService.findValues(text.orElse(""));
  }

  @GetMapping(value = "db/version", produces = MediaType.TEXT_PLAIN_VALUE)
  public String getVersion() {
    return this.carsService.getProcessedDate();
  }
}
